Adding a unit test suite

GenevieveBuckley opened the pull request:
Adds a unit test suite
Closes https://github.com/beeware/briefcase-template/issues/19

This PR:
- [x] Adds a `tox.ini` configuration file
- [x] Adds a GitHub Actions config to run CI tests
- [x] Tests that `pyproject.toml` can be parsed without error
- [x] Tests that there are no flake8 errors raised on any of the generated python code
- [x] Tests that all the generated python files can be compiled using [py_compile](https://docs.python.org/3/library/py_compile.html)
- [x] ~~Tests that the filename tree structure of the generated app matches what we expect~~

STILL TO DO:
- [x] ~~Expand to test all GUI frameworks (PySide, PursuedPyBear, etc), not just the default~~ Done!
- [x] ~~Replace `os.path` functions with `pathlib` where appropriate~~

I had assumed adding support for all the GUI frameworks would be as easy as adding a couple more fixtures and `@pytest.mark.parametrize` to all the tests -- but it seems like you can't pass fixtures into parametrize like that. Any suggestions?

freakboy3742 commented on tox.ini:
```diff
@@ -0,0 +1,11 @@
+  cookiecutter
+  flake8
+  pytest
+  tomlkit
```
TIL about tomlkit... I've been using vanilla `toml` in briefcase. Based on [this recommendation](https://twitter.com/uranusjr/status/1265812453599764481) from a PyPA member, I guess we should stick to toml, as we're only using the toml *parsing* capability, not the writing features.

GenevieveBuckley replied:
Oh, interesting! Thanks for sharing. I'd actually only recently come across tomlkit, but that was a context where I needed to write to file as well.

freakboy3742 commented on tox.ini:
```diff
@@ -0,0 +1,11 @@
+deps =
+  cookiecutter
+  flake8
+  pytest
```
BeeWare also uses the `pytest-tldr` plugin to make the `pytest` output more consumable (especially on test failure)

freakboy3742 commented on .github/workflows/run_tests.yml:
```diff
@@ -0,0 +1,22 @@
+    - name: Set up Python 3.8
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.8
```
As noted in the comments in tox.ini, we've found it's safest to run these sorts of tests on the "oldest supported Python" to make sure newer language idiom doesn't leak in.

freakboy3742 commented on pyproject.toml:
```diff
@@ -0,0 +1,3 @@
+[build-system]
```
This is only required because tox is (currently) configured to install the project; if you add `skip_install = True` to the `[testenv]` section of the Tox configuration, pyproject.toml won't be required.

freakboy3742 commented on tox.ini:
```diff
@@ -0,0 +1,11 @@
+[tox]
+envlist = py27, py37
+isolated_build = True
```
This is only needed because pyproject.toml exists.

GenevieveBuckley commented on .github/workflows/run_tests.yml:
```diff
@@ -0,0 +1,22 @@
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install tox
```
Good to know, thank you!

GenevieveBuckley commented on tests/test_app_template.py:
```diff
@@ -0,0 +1,110 @@
+def app_directory(tmpdir_factory):
+    """Fixture for a default app."""
+    output_dir = str(tmpdir_factory.mktemp("default-app"))
+    CCDS_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
```
Haha, I took pathlib out when I was trying to make the python2.7 tests pass! Good to know that's not something we need to worry about :)

GenevieveBuckley commented on tests/test_app_template.py:
```diff
@@ -0,0 +1,110 @@
+    yield output_dir
+
+    # cleanup after
+    shutil.rmtree(output_dir)
```
Ok, I'll switch it to a regular return statement. 
This fixture was modeled on one from the data science cookiecutter, so I assumed it was important.

GenevieveBuckley commented:
I've addressed all the points you made, thank you for the review!

Note that:
* I didn't add licenses other than the default to our test cases, primarily because all that changes is the contents of a single text file & there is no test checking that content currently. So as things stand now, it's a little meaningless.
* The fixture has a function-scope. Maybe that's slightly inefficient but on the whole it's fine because the 15 tests run in under 13 seconds. (That does bug me a little, I wanted to make it class scoped then pack the three tests into a single class & parameterize that. But I couldn't get that working nicely with the argument passing,)
